geomstats/stiefel.py

```diff
@@ -0,0 +1,279 @@
+"""
+Stiefel manifold St(p,n),
+a set of all orthonormal p-frames in n-dimensional space,
+where p <= n
+"""
+
+import logging
+import random
+
+import geomstats.backend as gs
+
+from geomstats.embedded_manifold import EmbeddedManifold
+from geomstats.riemannian_metric import RiemannianMetric
+
+TOLERANCE = 1e-6
+EPSILON = 1e-6
+
+def matrix_f(sq_mat, f):
+    
+    sq_mat = gs.to_ndarray(sq_mat, to_ndim=3)
+    
+    [eigenvalues, vectors] = gs.linalg.eig(sq_mat)
+    
+    exp_eigenvalues = f(eigenvalues)
+    
+    aux = gs.einsum('ijk,ik->ijk', vectors, exp_eigenvalues)
+    exp_mat = gs.einsum('ijk,ikl->ijl', aux, gs.linalg.inv(vectors))
+    
+    return exp_mat.real
+
+class Stiefel(EmbeddedManifold):
+
+    def __init__(self, p, n):
+        self.p = p
+        self.n = n
+
+        self.dimension = int(p * n - (p * (p - 1) / 2))
+
+    def belongs(self, point, tolerance=TOLERANCE):
+        """
+        Evaluate if a point belongs to St(p,n),
+        i.e. if it is a p-frame in n-dimensional space,
+        and it is orthonormal.
+        """
+        point = gs.to_ndarray(point, to_ndim=3)
+        (_, point_dim) = point.shape
+
+        if point_dim != (self.n, self.p):
+            return False
+        
+        point_norm = gs.norm(gs.dot(gs.transpose(point), point) - gs.eye(point_dim[1]))
+
+        return gs.less_equal(point_norm, tolerance)
+
+    def project(self):
+        pass
+
+    def random_uniform(self, n_samples=1):
+        """
+        Sample on St(p,n) with the uniform distribution.
+
+        If Z(p,n) ~ N(0,1), then St(p,n) ~ U, according to Haar measure:
+        St(p,n) := Z(Z^TZ)^{-1/2}
+        """
+        Z = gs.normal(size=(n_samples, self.n, self.p))
```

Comment:
avoid one letter variables

Reply:
while avoiding one letter variables is good practice overall, while dealing with matrices it's may be more readable to my taste

for example, variable names `Q, R` coming from QR-decomposition are speak for themselves, that they are orthogonal and upper-triangular matrices

moreover, I like namings `base_point`, `point`, `tangent_vector_a`, but for intermediate, auxiliary matrix-valued variables `Z` looks quite good for my opinion

what alternative name for this variable do you see?